index.d.ts

```diff
@@ -0,0 +1,438 @@
+import EmberObject from '@ember/object';
+import {
+  UnwrapComputedPropertyGetter,
+  UnwrapComputedPropertyGetters
+} from '@ember/object/-private/types';
+
+import RSVP from 'rsvp';
+
+// Lifted from @types/ember__object/observable.d.ts
+interface Getter {
+  /**
+   * Retrieves the value of a property from the object.
+   */
+  get<K extends keyof this>(key: K): UnwrapComputedPropertyGetter<this[K]>;
```

Comment:
This is going to break for cases like `this.get('foo.bar')`. Not all values that can be passed into this function are `keyof this`

Reply:
> Don't we just say getting in this way is not supported when using TS?

This is the current state of the types, given that we don't have the ability to use `unknown` as long as we continue to support TS 2.8.

Something like this (available in TS 3.0+) would support Ember's public API fully, without compromising type safety

```ts
interface Getter {
  get<K extends keyof this>(key: K): this[K];
  get(key: string): unknown;
}

interface Foo extends Getter {
  blue: string[];
  red: {
    green: string;
  }
}

const foo: Foo = {} as any;

foo.get('blue'); // string[];
foo.get('red.green'); // unknown
```

Reply:
We should note that as a point for discussion Soon(tm). 3.0 was a Ling time ago and the official support policy we decided on currently only commits us to 3.5 and 3.6, I believe. We’ve intentionally supported a wider range, but it’s been quite some time and it’s worth revisiting what we support (outside this PR!).